### ISSUE TITLE
Windows: Console event handler doesn't always work

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#5958]
We are installing a console control event handler on Windows through `SetConsoleCtrlHandler` to shut the server down when pressing CTRL+C among others.

This seems to work fine for CTRL+C, but when for example closing the console window, I'm instead getting the following message:
```
Caught signal. Waiting for initialization to complete before closing...
```

After a few seconds it closes the server. However I think it doesn't shut it down properly. At least none of the usual shutdown messages later appear in the server's log file, while with CTRL+C they end up there as expected.

I suspect the event might arrive on the wrong thread maybe?

@larkost first ran into this issue in the context of testing, where we need to use an event to shut down a server cleanly.

**Comment by AtnNn** [comment]
According to [the documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/ms683242(v=vs.85).aspx), pressing C-c or closing the console window should call the same handler, but with a different argument (`CTRL_C_EVENT` and `CTRL_CLOSE_EVENT` respectively).

The message about `Waiting for initialization` should only display if the thread pool hasn't been initialized yet. Does it appear if the console window is closed after the server becomes ready?

**Comment by danielmewes** [comment]
> Does it appear if the console window is closed after the server becomes ready?

Yes, the server had been ready and running for multiple seconds before I closed the window.

**Comment by AtnNn** [comment]
The bug seems to be that the event handler returns too early. With 04831d6 rethinkdb shuts down gracefully when the console window is closed.
